client/main.go

```diff
@@ -412,7 +412,224 @@ func discoverHTCondorToken(tokenName string) string {
 	return tokenLocation
 }
 
-// Start the transfer, whether read or write back
+func DoPut(sourceFile string, destination string, recursive bool) (bytesTransferred int64, err error) {
+	// First, create a handler for any panics that occur
+	defer func() {
+		if r := recover(); r != nil {
+			log.Debugln("Panic captured while attempting to perform transfer (DoPut):", r)
+			log.Debugln("Panic caused by the following", string(debug.Stack()))
+			ret := fmt.Sprintf("Unrecoverable error (panic) captured in DoPut: %v", r)
+			err = errors.New(ret)
+			bytesTransferred = 0
+
+			// Attempt to add the panic to the error accumulator
+			AddError(errors.New(ret))
+		}
+	}()
+
+	// Parse the source and destination with URL parse
+	sourceFile, source_scheme := correctURLWithUnderscore(sourceFile)
+	source_url, err := url.Parse(sourceFile)
+	if err != nil {
+		log.Errorln("Failed to parse source URL:", err)
+		return 0, err
+	}
+	source_url.Scheme = source_scheme
+
+	destination, dest_scheme := correctURLWithUnderscore(destination)
+	dest_url, err := url.Parse(destination)
+	if err != nil {
+		log.Errorln("Failed to parse destination URL:", err)
+		return 0, err
+	}
+	dest_url.Scheme = dest_scheme
+
+	if dest_url.Host != "" {
+		if dest_url.Scheme == "osdf" || dest_url.Scheme == "stash" {
+			dest_url.Path = "/" + path.Join(dest_url.Host, dest_url.Path)
+		} else if dest_url.Scheme == "pelican" {
+			federationUrl, _ := url.Parse(dest_url.String())
+			federationUrl.Scheme = "https"
+			federationUrl.Path = ""
+			viper.Set("Federation.DiscoveryUrl", federationUrl.String())
+			err = config.DiscoverFederation()
+			if err != nil {
+				return 0, err
+			}
+		}
+	}
+	destScheme, _ := getTokenName(dest_url)
+
+	understoodSchemes := []string{"file", "osdf", "pelican", ""}
```

Comment:
No, I asked Brian and he said that we should just worry about everything other than stash. Probably so we don't have to worry about backwards compatibility since stash can use `object copy`.